### DISCUSSION TITLE
プロジェクトにCDKサンプルを付けてほしい

### DISCUSSION THREAD

**Discussion by koichimurakami** [mbc-net/mbc-cqrs-serverless#34] [Ideas]
このフレームワークでローカルで開発できるが、AWSにデプロイする方法が分からない。
そのため、CDKのサンプルをテンプレートとして作成しておき、デプロイする際にはCDKサンプルを書き換えて使ってもらえるようにする。

ただし全てのAWS環境が同一でないためある程度AWS環境が構築されている前提としたい
以下については既に設定がされているものとする

* Route53
* ACM
* VPC
* RDS
* RDS接続設定 等

**Comment by koichimurakami** [comment]
事前に設定すべきネットワーク設定等はBlogで設定方法をお知らせする。

* ネットワーク設定: https://www.mbc-net.com/aws-%e7%92%b0%e5%a2%83%e6%a7%8b%e7%af%89-mbc-cqrs-%e3%82%b5%e3%83%bc%e3%83%90%e3%83%bc%e3%83%ac%e3%82%b9-%e3%83%95%e3%83%ac%e3%83%bc%e3%83%a0%e3%83%af%e3%83%bc%e3%82%af%e7%94%a8%e3%83%8d%e3%83%83/

**Reply by baquan1708** [reply]
@koichimurakami The following configurations are assumed to have already been set up.
- Route53
- ACM
- VPC
- RDS
- RDS connection settings.
- email from the Simple Email Service
- DynamoDB table name

Does the CDK sample we provide support deployment only to Lambda, or does it also support ECS when a user creates a new project with the `mbc new` command?

**Reply by koichimurakami** [reply]
@baquan1708 
it's better that we could select Lambda or ECS.
Is it possible to select it in the config?

**Reply by baquan1708** [reply]
Yes, it's possible to select this in the config. I just wanted to confirm how limited the CDK sample should be. 
I'll go ahead and create the CDK sample now.

**Reply by koichimurakami** [reply]
@baquan1708 
Should ECS instance be behind of ALB/API Gateway?
![CDK_Infrastructure](https://github.com/user-attachments/assets/da0c2fa4-0dc8-431e-8986-3d3ea820d819)

I think, its better that dynamodb streams should kick lambda instead of ecs

**Reply by baquan1708** [reply]
@koichimurakami You're absolutely correct. We will place the ECS instances behind an ALB or API Gateway, and Lambda will handle DynamoDB streams and SQS processing instead of ECS.

In the architecture depicted above, there will be a connection flow as follows: CloudFront → API Gateway → ALB → ECS.
![image](https://github.com/user-attachments/assets/1af1d565-13bd-4fad-8342-b1b1fae3955c)

**Reply by baquan1708** [reply]
@koichimurakami I've just created PR https://github.com/mbc-net/mbc-cqrs-serverless/pull/40: add cdk sample to template. Could you please review it?

cc: @Tuan-Anh-Nguyen-pionero

**Reply by baquan1708** [reply]
@koichimurakami How to use cdk infra sample in infra folder to deploy to AWS env.

```txt
## Deployment

The infrastructure that inside infra folder to deploy to AWS environment. The cdk pure support deploy to separate `dev`, `staging` and `production` environment.

The following configurations are assumed to have been set up:

- Route53
- ACM
- VPC
- RDS
- RDS connection settings.
- email from the Simple Email Service
- DynamoDB table name
- Code connection
- WAF (optional)

With these in place, modify the content inside `infra/config/constant.ts`. For dev environment, update the files `infra/config/dev/index.ts`. Similarly, you can update `infra/config/stg/index.ts` for the staging environment and `infra/config/prod/index.ts` for the production environment.

Next, update `infra/bin/infra.ts` with the AWS account and region you want to deploy to.

To deploy, first run `npx cdk bootstrap` to bootstrap the CDK environment. Then, execute `npx cdk deploy` to deploy.
```

**Reply by koichimurakami** [reply]
@baquan1708 
Thank you so much. 
I will check it.

**Comment by Tuan-Anh-Nguyen-pionero** [comment]
@baquan1708 cc @koichimurakami 
Could you implement cdk snapshot test and add it to github action. When user create PR, run test.
Here is spec: with random value in this InfraStack -> run test with resuit is PASS.
Pls show your solution in this Discussion.

**Reply by baquan1708** [reply]
@Tuan-Anh-Nguyen-pionero  Yes, we could add a snapshot test job to the workflow in GitHub Actions and run the workflow when a user creates a PR.

As you discussed, there is a problem: when we run the infrastack snapshot test, some values (e.g., `expiration time`, `header values`, etc.) in the CDK template may differ each time the test is run, which could cause the test to fail.

I explored and tried out three methods to handle this issue:

1. Pass key/value pairs to the `toMatchSnapshot` method to change the snapshot file (Reference: [Property Matcher](https://jestjs.io/docs/snapshot-testing#property-matchers)).
2. Modify the value to be compared before calling `toMatchSnapshot` (Reference: [Property Matcher](https://jestjs.io/docs/snapshot-testing#property-matchers) tip section).
3. Mock the function responsible for generating random values (Reference: [Mock Module](https://jestjs.io/docs/mock-functions#mocking-modules)).

Additionally, there's another option: using a [snapshot serializer](https://jestjs.io/docs/configuration#snapshotserializers-arraystring), which can change how snapshot files are read/write. This provides strong support for modifying snapshot content during comparisons.

**Note**: If the snapshot file has many properties, Methods 1, 2, and using a snapshot serializer require careful adjustments to ensure the values/keys are updated appropriately.
